File: TS/系统/03．技能系统/00．技能模板+函数/04．机制组件/09．装备通用机制/04．失去资源属性转换.ts
/** @noSelfInFile */

const jass = require("jass.common") as any;

const GetUnitState = jass.GetUnitState as (whichUnit: any, whichUnitState: any) => number;
const IsUnitType = jass.IsUnitType as (whichUnit: any, whichUnitType: any) => boolean;
const R2I = jass.R2I as (value: number) => number;
const UNIT_STATE_LIFE = jass.UNIT_STATE_LIFE as any;
const UNIT_STATE_MANA = jass.UNIT_STATE_MANA as any;
const UNIT_STATE_MAX_LIFE = jass.UNIT_STATE_MAX_LIFE as any;
const UNIT_STATE_MAX_MANA = jass.UNIT_STATE_MAX_MANA as any;
const UNIT_TYPE_DEAD = jass.UNIT_TYPE_DEAD as any;

const { addPeriodicCallback, removePeriodicCallback } = require("系统.00．核心系统.05．中心计时器") as {
  addPeriodicCallback: (this: void, intervalMs: number, callback: (this: void) => void) => number;
  removePeriodicCallback: (this: void, id: number) => void;
};

export type 失去资源类型 = "生命" | "魔法";

export interface 失去资源属性事件 {
  单位: any;
  旧档位: number;
  新档位: number;
  缺失比例: number;
  满档比例: number;
}

export interface 失去资源属性转换参数 {
  名称?: string;
  单位: any;
  资源类型: 失去资源类型;
  满档缺失比例: number;
  档位数量?: number;
  检查间隔毫秒?: number;
  on档位变化: (this: void, event: 失去资源属性事件) => void;
}

export interface 失去资源属性转换控制器 {
  readonly 名称: string;
  读取档位(): number;
  刷新(): number;
  停止(): void;
}

function 单位有效(this: void, 单位: any): boolean {
  return 单位 != null && 单位 !== 0 && IsUnitType(单位, UNIT_TYPE_DEAD) !== true;
}

function 取资源比例(this: void, 单位: any, 类型: 失去资源类型): number {
  const current = GetUnitState(单位, 类型 === "生命" ? UNIT_STATE_LIFE : UNIT_STATE_MANA);
  const max = GetUnitState(单位, 类型 === "生命" ? UNIT_STATE_MAX_LIFE : UNIT_STATE_MAX_MANA);
  if (max <= 0) return 0;
  const missing = max - current;
  if (missing <= 0) return 0;
  return missing / max;
}

class 失去资源属性转换实现 implements 失去资源属性转换控制器 {
  readonly 名称: string;
  readonly 控制器ID: number;
  private 参数: 失去资源属性转换参数;
  private 当前档位 = 0;
  private 已停止 = false;

  constructor(名称: string, 参数: 失去资源属性转换参数, 控制器ID: number) {
    this.名称 = 名称;
    this.参数 = 参数;
    this.控制器ID = 控制器ID;
    this.刷新();
    失去资源属性转换表[this.控制器ID] = this;
    确保失去资源属性转换Tick(参数.检查间隔毫秒 ?? 200);
  }

  读取档位(): number {
    return this.当前档位;
  }

  刷新(): number {
    if (this.已停止) return this.当前档位;
    if (!单位有效(this.参数.单位)) {
      this.设置档位(0, 0);
      return this.当前档位;
    }
    const ratio = 取资源比例(this.参数.单位, this.参数.资源类型);
    const maxRatio = this.参数.满档缺失比例 > 0 ? this.参数.满档缺失比例 : 1;
    let normalized = ratio / maxRatio;
    if (normalized < 0) normalized = 0;
    if (normalized > 1) normalized = 1;
    const 档位数量 = this.参数.档位数量 != null && this.参数.档位数量 > 0 ? this.参数.档位数量 : 100;
    const 新档位 = R2I(normalized * 档位数量);
    this.设置档位(新档位, ratio);
    return this.当前档位;
  }

  停止(): void {
    if (this.已停止) return;
    this.已停止 = true;
    delete 失去资源属性转换表[this.控制器ID];
    尝试停止失去资源属性转换Tick();
    if (this.当前档位 !== 0) {
      this.设置档位(0, 0);
    }
  }

  private 设置档位(新档位: number, 缺失比例: number): void {
    if (新档位 === this.当前档位) return;
    const 旧档位 = this.当前档位;
    this.当前档位 = 新档位;
    this.参数.on档位变化({ 单位: this.参数.单位, 旧档位, 新档位, 缺失比例, 满档比例: this.参数.满档缺失比例 });
  }
}

const 失去资源属性转换表: Record<number, 失去资源属性转换实现> = {};
let 失去资源属性转换计数 = 0;
let 失去资源属性转换TickID = 0;

function 确保失去资源属性转换Tick(this: void, interval: number): void {
  if (失去资源属性转换TickID !== 0) return;
  失去资源属性转换TickID = addPeriodicCallback(interval, on失去资源属性转换Tick);
}

function 尝试停止失去资源属性转换Tick(this: void): void {
  for (const key in 失去资源属性转换表) {
    if (失去资源属性转换表[key] != null) return;
  }
  if (失去资源属性转换TickID !== 0) {
    removePeriodicCallback(失去资源属性转换TickID);
    失去资源属性转换TickID = 0;
  }
}

export function 创建失去资源属性转换(this: void, 参数: 失去资源属性转换参数): 失去资源属性转换控制器 {
  失去资源属性转换计数 += 1;
  return new 失去资源属性转换实现(参数.名称 ?? "失去资源属性转换", 参数, 失去资源属性转换计数);
}

function on失去资源属性转换Tick(this: void): void {
  for (const key in 失去资源属性转换表) {
    const 控制器 = 失去资源属性转换表[key];
    if (控制器 != null) 控制器.刷新();
  }
}

File: src/系统/03．技能系统/00．技能模板+函数/04．机制组件/09．装备通用机制/04．失去资源属性转换.lua
local ____lualib = require("lualib_bundle")
local __TS__Class = ____lualib.__TS__Class
local __TS__New = ____lualib.__TS__New
local ____exports = {}
local ____on_5931_53BB_8D44_6E90_5C5E_6027_8F6C_6362Tick
function ____on_5931_53BB_8D44_6E90_5C5E_6027_8F6C_6362Tick(variable)
    local _____63A7_5236_5668 = variable
    if _____63A7_5236_5668 ~= nil then
        _____63A7_5236_5668["刷新"](_____63A7_5236_5668)
    end
end
---
-- @noSelfInFile
local jass = require("jass.common")
local japi = require("jass.japi")
local GetUnitStateJapi = japi.GetUnitState
local GetUnitState = jass.GetUnitState
local IsUnitType = jass.IsUnitType
local R2I = jass.R2I
local UNIT_STATE_LIFE = jass.UNIT_STATE_LIFE
local UNIT_STATE_MANA = jass.UNIT_STATE_MANA
local UNIT_STATE_MAX_LIFE = jass.UNIT_STATE_MAX_LIFE
local UNIT_STATE_MAX_MANA = jass.UNIT_STATE_MAX_MANA
local UNIT_TYPE_DEAD = jass.UNIT_TYPE_DEAD
local ____require_result_0 = require("系统.00．核心系统.05．中心计时器")
local addPeriodicCallback = ____require_result_0.addPeriodicCallback
local removePeriodicCallback = ____require_result_0.removePeriodicCallback
local function _____5355_4F4D_6709_6548(_____5355_4F4D)
    return _____5355_4F4D ~= nil and _____5355_4F4D ~= 0 and IsUnitType(_____5355_4F4D, UNIT_TYPE_DEAD) ~= true
end
local function _____53D6_8D44_6E90_6BD4_4F8B(_____5355_4F4D, _____7C7B_578B)
    local ____5355_4F4D_2 = _____5355_4F4D
    local ____temp_1
    if _____7C7B_578B == "生命" then
        ____temp_1 = UNIT_STATE_LIFE
    else
        ____temp_1 = UNIT_STATE_MANA
    end
    local current = GetUnitState(____5355_4F4D_2, ____temp_1)
    local ____5355_4F4D_4 = _____5355_4F4D
    local ____temp_3
    if _____7C7B_578B == "生命" then
        ____temp_3 = UNIT_STATE_MAX_LIFE
    else
        ____temp_3 = UNIT_STATE_MAX_MANA
    end
    local max = GetUnitStateJapi(____5355_4F4D_4, ____temp_3)
    if max <= 0 then
        return 0
    end
    local missing = max - current
    if missing <= 0 then
        return 0
    end
    return missing / max
end
local _____5931_53BB_8D44_6E90_5C5E_6027_8F6C_6362_5B9E_73B0 = __TS__Class()
_____5931_53BB_8D44_6E90_5C5E_6027_8F6C_6362_5B9E_73B0.name = "失去资源属性转换实现"
function _____5931_53BB_8D44_6E90_5C5E_6027_8F6C_6362_5B9E_73B0.prototype.____constructor(self, _____540D_79F0, _____53C2_6570, _____63A7_5236_5668ID)
    self["当前档位"] = 0
    self["Tick回调ID"] = 0
    self["已停止"] = false
    self["名称"] = _____540D_79F0
    self["参数"] = _____53C2_6570
    self["控制器ID"] = _____63A7_5236_5668ID
    self["刷新"](self)
    self["Tick回调ID"] = addPeriodicCallback(_____53C2_6570["检查间隔毫秒"] or 200, ____on_5931_53BB_8D44_6E90_5C5E_6027_8F6C_6362Tick, self)
end
_____5931_53BB_8D44_6E90_5C5E_6027_8F6C_6362_5B9E_73B0.prototype["读取档位"] = function(self)
    return self["当前档位"]
end
_____5931_53BB_8D44_6E90_5C5E_6027_8F6C_6362_5B9E_73B0.prototype["刷新"] = function(self)
    if self["已停止"] then
        return self["当前档位"]
    end
    if not _____5355_4F4D_6709_6548(self["参数"]["单位"]) then
        self["设置档位"](self, 0, 0)
        return self["当前档位"]
    end
    local ratio = _____53D6_8D44_6E90_6BD4_4F8B(self["参数"]["单位"], self["参数"]["资源类型"])
    local maxRatio = self["参数"]["满档缺失比例"] > 0 and self["参数"]["满档缺失比例"] or 1
    local normalized = ratio / maxRatio
    if normalized < 0 then
        normalized = 0
    end
    if normalized > 1 then
        normalized = 1
    end
    local _____6863_4F4D_6570_91CF = self["参数"]["档位数量"] ~= nil and self["参数"]["档位数量"] > 0 and self["参数"]["档位数量"] or 100
    local _____65B0_6863_4F4D = R2I(normalized * _____6863_4F4D_6570_91CF)
    self["设置档位"](self, _____65B0_6863_4F4D, ratio)
    return self["当前档位"]
end
_____5931_53BB_8D44_6E90_5C5E_6027_8F6C_6362_5B9E_73B0.prototype["停止"] = function(self)
    if self["已停止"] then
        return
    end
    self["已停止"] = true
    if self["Tick回调ID"] ~= 0 then
        removePeriodicCallback(self["Tick回调ID"])
        self["Tick回调ID"] = 0
    end
    if self["当前档位"] ~= 0 then
        self["设置档位"](self, 0, 0)
    end
end
_____5931_53BB_8D44_6E90_5C5E_6027_8F6C_6362_5B9E_73B0.prototype["设置档位"] = function(self, _____65B0_6863_4F4D, _____7F3A_5931_6BD4_4F8B)
    if _____65B0_6863_4F4D == self["当前档位"] then
        return
    end
    local _____65E7_6863_4F4D = self["当前档位"]
    self["当前档位"] = _____65B0_6863_4F4D
    self["参数"]["on档位变化"]({
        ["单位"] = self["参数"]["单位"],
        ["旧档位"] = _____65E7_6863_4F4D,
        ["新档位"] = _____65B0_6863_4F4D,
        ["缺失比例"] = _____7F3A_5931_6BD4_4F8B,
        ["满档比例"] = self["参数"]["满档缺失比例"]
    })
end
local _____5931_53BB_8D44_6E90_5C5E_6027_8F6C_6362_8BA1_6570 = 0
____exports["创建失去资源属性转换"] = function(_____53C2_6570)
    _____5931_53BB_8D44_6E90_5C5E_6027_8F6C_6362_8BA1_6570 = _____5931_53BB_8D44_6E90_5C5E_6027_8F6C_6362_8BA1_6570 + 1
    return __TS__New(_____5931_53BB_8D44_6E90_5C5E_6027_8F6C_6362_5B9E_73B0, _____53C2_6570["名称"] or "失去资源属性转换", _____53C2_6570, _____5931_53BB_8D44_6E90_5C5E_6027_8F6C_6362_8BA1_6570)
end
return ____exports
